Tag cannot be a empty string

miaoyq opened the pull request:
Tag cannot be a empty string, because it will be used in the following code。

liggitt commented on pkg/image/registry/imagestreamimport/rest.go:
```diff
@@ -305,9 +305,11 @@ func checkImportFailure(status api.ImageImportStatus, stream *api.ImageStream, t
-				tag = ref.Tag
+			ref, err := api.ParseDockerImageReference(status.Image.DockerImageReference)
+			if err != nil || ref.Tag == "" {
+				return false
```
need a test exercising this? it's not super clear what returning false from this function means

miaoyq replied:
Thanks for review! @liggitt 
`api.HasTagCondition` and `api.SetTagCondition` are meaningless, if `tag` is an empty string.
So we have to ensure that `tag` is a non empty string.

mfojtik commented:
@miaoyq I agree with @liggitt. Can you add a simple unit test exercising this?

miminar commented on pkg/image/registry/imagestreamimport/rest.go:
```diff
@@ -305,9 +305,11 @@ func checkImportFailure(status api.ImageImportStatus, stream *api.ImageStream, t
-			if ref, err := api.ParseDockerImageReference(status.Image.DockerImageReference); err == nil {
-				tag = ref.Tag
+			ref, err := api.ParseDockerImageReference(status.Image.DockerImageReference)
+			if err != nil || ref.Tag == "" {
```
*nitpick*: `len(ref.Tag) == 0`

miaoyq commented:
@liggitt @mfojtik @miminar Thanks for review! First up, my most sincere apologies for the delay here,  have been out of commission for a while and was just catching up. 
I have updated the function and  added a unit test for it.  PTLK, Thanks!

miminar commented on pkg/image/registry/imagestreamimport/rest.go:
```diff
@@ -305,13 +305,14 @@ func checkImportFailure(status api.ImageImportStatus, stream *api.ImageStream, t
 	}
 
-	if !api.HasTagCondition(stream, tag, condition) {
+	if tag != "" && !api.HasTagCondition(stream, tag, condition) {
```
*nit*: `len(tag) > 0`

miminar commented on pkg/image/registry/imagestreamimport/rest_test.go:
```diff
@@ -198,3 +198,131 @@ func TestImportSuccessful(t *testing.T) {
+		stream         *api.ImageStream
+		tag            string
+		nextGeneration int64
+		now            unversioned.Time
```
*nit*: why not use constants instead of two attributes above? They seem not to be overridden below.

miaoyq replied:
Updated.  PTAL Thanks! @miminar

miminar replied:
Thanks, LGTM.

miminar commented on pkg/image/registry/imagestreamimport/rest.go:
```diff
@@ -305,13 +305,14 @@ func checkImportFailure(status api.ImageImportStatus, stream *api.ImageStream, t
 		} else if status.Image != nil {
-			if ref, err := api.ParseDockerImageReference(status.Image.DockerImageReference); err == nil {
+			ref, err := api.ParseDockerImageReference(status.Image.DockerImageReference)
+			if err == nil {
```
There's no need for this change. We actually prefer the original way of handling errors.

miminar commented:
[test]

miminar commented on pkg/image/registry/imagestreamimport/rest.go:
```diff
@@ -311,7 +311,7 @@ func checkImportFailure(status api.ImageImportStatus, stream *api.ImageStream, t
 	}
 
-	if !api.HasTagCondition(stream, tag, condition) {
+	if len(tag) > 0 && !api.HasTagCondition(stream, tag, condition) {
```
I'm sorry, one more thing. The conditions will be lost if we don't set them. Could you please log them in some nice way with ERROR level for `len(tag) == 0 && !api.HasTagCondition(stream, tag, condition)`?

miaoyq replied:
updated, PTAL Thanks! @miminar

miminar commented:
Flake #11240 in job https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/9801/. re-[test]

miminar commented on pkg/image/registry/imagestreamimport/rest.go:
```diff
@@ -312,11 +312,15 @@ func checkImportFailure(status api.ImageImportStatus, stream *api.ImageStream, t
+				stream.Spec.Tags[tag] = tagRef
+			}
+		} else {
+			glog.Errorln("tag conditions can't be set, because the tag of image is empty string")
```
Could you rewrite to something like:

```go
glog.Error("tag unset - not setting failed import condition on stream=\"%s/%s\" reason=%q: %s", stream.Namespace, stream.Name, condition.Reason, condition.Message)
```

miaoyq replied:
I have updated.
Thank you for your patience also! :smile:

miminar commented:
Thanks! re-[test]

miminar commented:
Flake #8571 in job https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/9887/. re-[test]

miminar commented:
Flake #8571 once again in job https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/9948/.

@mfojtik I believe this is good to *merge*.

miaoyq commented:
@mfojtik  /cc   Thanks!

miaoyq commented:
@mfojtik PTAL, Thanks so much!

mfojtik commented:
[test]

miminar commented:
@miaoyq sorry, for the inactivity; would you mind doing a rebase?
